### DISCUSSION TITLE
PreserveReferences uses GetHashCode?

### DISCUSSION THREAD

**Discussion by IB38** [LuckyPennySoftware/AutoMapper#3654] [NA]
Hi,

Does automapper use GetHashCode for unique object identification?
I've run into some unexpected behavior when using automapper to map classes with overriden GetHashCode with PreserveReferences option.
Perhaps using RuntimeHelpers.GetHashCode would be a better option?

**Comment by lbargaoanu** [comment]
Other than the custom `GetHashCode` being broken, why wouldn't we call it virtually?

**Reply by IB38** [reply]
Using GetHashCode for object identification is not recommended, but i dont think there is any other way other than using Object.GetHashCode.

Object.GetHashCode result being based on reference is just an effect of implementation. When GetHashCode and Equals are overriden, often its done to use property/field/etc equality instead of reference equality. In this case, AutoMapper calling GetHashCode virtually is incorrent, because the intent is to get some representation of reference to use as key in object dictionary for PreserveReferences. Something like that.

**Reply by lbargaoanu** [reply]
If you override `Equals`, that seems to me like a reasonable result and some people might rely on that, but I see your point too.
